File: src/evaluate-condition.test.ts
import { describe, expect, it, vi, beforeEach, afterEach } from 'vitest';
import mockFs from 'mock-fs';
import { evaluateCondition } from './evaluate-condition';
import type { Condition } from './types';

describe('evaluateCondition', () => {
  beforeEach(() => {
    vi.unstubAllEnvs();
  });

  afterEach(() => {
    vi.unstubAllEnvs();
    mockFs.restore();
  });

  describe('env_set', () => {
    it('is true when the variable has a non-empty value', async () => {
      vi.stubEnv('SOME_VAR', '1');
      expect(
        await evaluateCondition({ type: 'env_set', name: 'SOME_VAR' })
      ).toBe(true);
    });

    it('is false when the variable is unset', async () => {
      vi.stubEnv('SOME_VAR', '');
      expect(
        await evaluateCondition({ type: 'env_set', name: 'SOME_VAR' })
      ).toBe(false);
    });

    it('is false when the variable is set to an empty string', async () => {
      // `env_set` uses Boolean(), so an empty string is falsy.
      vi.stubEnv('SOME_VAR', '');
      expect(
        await evaluateCondition({ type: 'env_set', name: 'SOME_VAR' })
      ).toBe(false);
    });
  });

  describe('env_value', () => {
    it('is true when the value matches exactly', async () => {
      vi.stubEnv('ROLE', 'agent-exec');
      expect(
        await evaluateCondition({
          type: 'env_value',
          name: 'ROLE',
          value: 'agent-exec',
        })
      ).toBe(true);
    });

    it('is false when the value differs', async () => {
      vi.stubEnv('ROLE', 'something-else');
      expect(
        await evaluateCondition({
          type: 'env_value',
          name: 'ROLE',
          value: 'agent-exec',
        })
      ).toBe(false);
    });

    it('is false when the variable is unset', async () => {
      expect(
        await evaluateCondition({
          type: 'env_value',
          name: 'ROLE',
          value: 'agent-exec',
        })
      ).toBe(false);
    });

    it('does not treat an empty target value as matching an unset variable', async () => {
      // process.env access for an unset var is undefined, not '', so an
      // `env_value` looking for '' must not match.
      expect(
        await evaluateCondition({ type: 'env_value', name: 'ROLE', value: '' })
      ).toBe(false);
    });
  });

  describe('file_exists', () => {
    it('is true when the path exists', async () => {
      mockFs({ '/opt/.devin': mockFs.directory({ mode: 0o755 }) });
      expect(
        await evaluateCondition({ type: 'file_exists', path: '/opt/.devin' })
      ).toBe(true);
    });

    it('is false when the path does not exist', async () => {
      mockFs({});
      expect(
        await evaluateCondition({ type: 'file_exists', path: '/opt/.devin' })
      ).toBe(false);
    });

    it('is false when access throws (e.g. unreadable parent)', async () => {
      mockFs({ '/opt': mockFs.directory({ mode: 0o000 }) });
      expect(
        await evaluateCondition({ type: 'file_exists', path: '/opt/.devin' })
      ).toBe(false);
    });
  });

  describe('env_matches', () => {
    it('is true when the value matches the pattern', async () => {
      vi.stubEnv('PATH', '/usr/bin:/home/me/.pi/agent/bin');
      expect(
        await evaluateCondition({
          type: 'env_matches',
          name: 'PATH',
          pattern: '\\.pi[\\\\/]agent',
        })
      ).toBe(true);
    });

    it('matches a backslash path separator too', async () => {
      vi.stubEnv('PATH', 'C:\\Users\\me\\.pi\\agent\\bin');
      expect(
        await evaluateCondition({
          type: 'env_matches',
          name: 'PATH',
          pattern: '\\.pi[\\\\/]agent',
        })
      ).toBe(true);
    });

    it('is false when the value does not match', async () => {
      vi.stubEnv('PATH', '/usr/bin:/usr/local/bin');
      expect(
        await evaluateCondition({
          type: 'env_matches',
          name: 'PATH',
          pattern: '\\.pi[\\\\/]agent',
        })
      ).toBe(false);
    });

    it('is false when the variable is unset', async () => {
      vi.stubEnv('TERM_PROGRAM', '');
      expect(
        await evaluateCondition({
          type: 'env_matches',
          name: 'TERM_PROGRAM',
          pattern: 'kiro',
        })
      ).toBe(false);
    });

    it('is false when the pattern is invalid rather than throwing', async () => {
      vi.stubEnv('TERM_PROGRAM', 'kiro');
      expect(
        await evaluateCondition({
          type: 'env_matches',
          name: 'TERM_PROGRAM',
          pattern: '(',
        })
      ).toBe(false);
    });
  });

  describe('no_tty', () => {
    const original = process.stdout.isTTY;
    afterEach(() => {
      process.stdout.isTTY = original;
    });

    it('is true when stdout is not a TTY', async () => {
      process.stdout.isTTY = false;
      expect(await evaluateCondition({ type: 'no_tty' })).toBe(true);
    });

    it('is false when stdout is a TTY', async () => {
      process.stdout.isTTY = true;
      expect(await evaluateCondition({ type: 'no_tty' })).toBe(false);
    });
  });

  describe('anyOf', () => {
    it('is true when at least one child is true', async () => {
      vi.stubEnv('B', '1');
      expect(
        await evaluateCondition({
          type: 'anyOf',
          conditions: [
            { type: 'env_set', name: 'A' },
            { type: 'env_set', name: 'B' },
          ],
        })
      ).toBe(true);
    });

    it('is false when all children are false', async () => {
      expect(
        await evaluateCondition({
          type: 'anyOf',
          conditions: [
            { type: 'env_set', name: 'A' },
            { type: 'env_set', name: 'B' },
          ],
        })
      ).toBe(false);
    });

    it('is false for an empty condition list', async () => {
      expect(await evaluateCondition({ type: 'anyOf', conditions: [] })).toBe(
        false
      );
    });

    it('is true when a later child is satisfied but an earlier one is not', async () => {
      vi.stubEnv('A', '');
      vi.stubEnv('B', '1');
      expect(
        await evaluateCondition({
          type: 'anyOf',
          conditions: [
            { type: 'env_set', name: 'A' },
            { type: 'env_set', name: 'B' },
          ],
        })
      ).toBe(true);
    });
  });

  describe('allOf', () => {
    it('is true when all children are true', async () => {
      vi.stubEnv('A', '1');
      vi.stubEnv('B', '1');
      expect(
        await evaluateCondition({
          type: 'allOf',
          conditions: [
            { type: 'env_set', name: 'A' },
            { type: 'env_set', name: 'B' },
          ],
        })
      ).toBe(true);
    });

    it('is false when any child is false', async () => {
      vi.stubEnv('A', '1');
      expect(
        await evaluateCondition({
          type: 'allOf',
          conditions: [
            { type: 'env_set', name: 'A' },
            { type: 'env_set', name: 'B' },
          ],
        })
      ).toBe(false);
    });

    it('is true for an empty condition list (vacuous truth)', async () => {
      expect(await evaluateCondition({ type: 'allOf', conditions: [] })).toBe(
        true
      );
    });

    it('is false when a later child is unsatisfied but an earlier one is not', async () => {
      vi.stubEnv('A', '1');
      vi.stubEnv('B', '');
      expect(
        await evaluateCondition({
          type: 'allOf',
          conditions: [
            { type: 'env_set', name: 'A' },
            { type: 'env_set', name: 'B' },
          ],
        })
      ).toBe(false);
    });
  });

  describe('nesting', () => {
    it('evaluates the cowork-style allOf(env_set, anyOf(...)) tree', async () => {
      vi.stubEnv('CLAUDE_CODE_IS_COWORK', '1');
      vi.stubEnv('CLAUDECODE', '1');
      const condition: Condition = {
        type: 'allOf',
        conditions: [
          { type: 'env_set', name: 'CLAUDE_CODE_IS_COWORK' },
          {
            type: 'anyOf',
            conditions: [
              { type: 'env_set', name: 'CLAUDECODE' },
              { type: 'env_set', name: 'CLAUDE_CODE' },
            ],
          },
        ],
      };
      expect(await evaluateCondition(condition)).toBe(true);
    });

    it('is false when the nested anyOf has no satisfied child', async () => {
      vi.stubEnv('CLAUDE_CODE_IS_COWORK', '1');
      vi.stubEnv('CLAUDECODE', '');
      vi.stubEnv('CLAUDE_CODE', '');
      const condition: Condition = {
        type: 'allOf',
        conditions: [
          { type: 'env_set', name: 'CLAUDE_CODE_IS_COWORK' },
          {
            type: 'anyOf',
            conditions: [
              { type: 'env_set', name: 'CLAUDECODE' },
              { type: 'env_set', name: 'CLAUDE_CODE' },
            ],
          },
        ],
      };
      expect(await evaluateCondition(condition)).toBe(false);
    });
  });

  describe('unknown condition type', () => {
    it('returns false for an unrecognized type', async () => {
      // Simulates a condition type not covered by the switch — the `default`
      // branch must fail closed rather than throw.
      const unknown = { type: 'not_a_real_type' } as unknown as Condition;
      expect(await evaluateCondition(unknown)).toBe(false);
    });
  });
});

File: evaluate_condition_test.go
package detectagent

import (
	"os"
	"testing"
)

func TestEvaluateCondition_EnvSet(t *testing.T) {
	t.Run("true when variable has a non-empty value", func(t *testing.T) {
		t.Setenv("SOME_VAR", "1")
		got, err := EvaluateCondition(Condition{Type: "env_set", Name: "SOME_VAR"})
		assertBool(t, err, got, true)
	})

	t.Run("false when variable is unset", func(t *testing.T) {
		os.Unsetenv("SOME_VAR")
		got, err := EvaluateCondition(Condition{Type: "env_set", Name: "SOME_VAR"})
		assertBool(t, err, got, false)
	})

	t.Run("false when variable is set to empty string", func(t *testing.T) {
		t.Setenv("SOME_VAR", "")
		got, err := EvaluateCondition(Condition{Type: "env_set", Name: "SOME_VAR"})
		assertBool(t, err, got, false)
	})
}

func TestEvaluateCondition_EnvValue(t *testing.T) {
	t.Run("true when value matches exactly", func(t *testing.T) {
		t.Setenv("ROLE", "agent-exec")
		got, err := EvaluateCondition(Condition{Type: "env_value", Name: "ROLE", Value: "agent-exec"})
		assertBool(t, err, got, true)
	})

	t.Run("false when value differs", func(t *testing.T) {
		t.Setenv("ROLE", "something-else")
		got, err := EvaluateCondition(Condition{Type: "env_value", Name: "ROLE", Value: "agent-exec"})
		assertBool(t, err, got, false)
	})

	t.Run("false when variable is unset", func(t *testing.T) {
		os.Unsetenv("ROLE")
		got, err := EvaluateCondition(Condition{Type: "env_value", Name: "ROLE", Value: "agent-exec"})
		assertBool(t, err, got, false)
	})

	t.Run("does not treat empty target as matching unset variable", func(t *testing.T) {
		// os.LookupEnv distinguishes unset from set-to-empty; an unset var
		// must not match value "".
		os.Unsetenv("ROLE")
		got, err := EvaluateCondition(Condition{Type: "env_value", Name: "ROLE", Value: ""})
		assertBool(t, err, got, false)
	})
}

func TestEvaluateCondition_FileExists(t *testing.T) {
	t.Run("true when path exists", func(t *testing.T) {
		dir := t.TempDir()
		got, err := EvaluateCondition(Condition{Type: "file_exists", Path: dir})
		assertBool(t, err, got, true)
	})

	t.Run("false when path does not exist", func(t *testing.T) {
		got, err := EvaluateCondition(Condition{Type: "file_exists", Path: "/nonexistent/path/abc123"})
		assertBool(t, err, got, false)
	})
}

func TestEvaluateCondition_EnvMatches(t *testing.T) {
	t.Run("true when value matches the pattern", func(t *testing.T) {
		t.Setenv("PATH", `/usr/bin:/home/me/.pi/agent/bin`)
		got, err := EvaluateCondition(Condition{Type: "env_matches", Name: "PATH", Pattern: `\.pi[\\/]agent`})
		assertBool(t, err, got, true)
	})

	t.Run("matches a backslash path separator too", func(t *testing.T) {
		t.Setenv("PATH", `C:\Users\me\.pi\agent\bin`)
		got, err := EvaluateCondition(Condition{Type: "env_matches", Name: "PATH", Pattern: `\.pi[\\/]agent`})
		assertBool(t, err, got, true)
	})

	t.Run("false when value does not match", func(t *testing.T) {
		t.Setenv("PATH", "/usr/bin:/usr/local/bin")
		got, err := EvaluateCondition(Condition{Type: "env_matches", Name: "PATH", Pattern: `\.pi[\\/]agent`})
		assertBool(t, err, got, false)
	})

	t.Run("false when variable is unset or empty", func(t *testing.T) {
		t.Setenv("TERM_PROGRAM", "")
		got, err := EvaluateCondition(Condition{Type: "env_matches", Name: "TERM_PROGRAM", Pattern: "kiro"})
		assertBool(t, err, got, false)
	})

	t.Run("false when pattern is invalid rather than erroring", func(t *testing.T) {
		t.Setenv("TERM_PROGRAM", "kiro")
		got, err := EvaluateCondition(Condition{Type: "env_matches", Name: "TERM_PROGRAM", Pattern: "("})
		assertBool(t, err, got, false)
	})
}

func TestEvaluateCondition_NoTTY(t *testing.T) {
	t.Run("true when stdout is not a TTY", func(t *testing.T) {
		orig := isTTYFn
		isTTYFn = func() bool { return false }
		t.Cleanup(func() { isTTYFn = orig })

		got, err := EvaluateCondition(Condition{Type: "no_tty"})
		assertBool(t, err, got, true)
	})

	t.Run("false when stdout is a TTY", func(t *testing.T) {
		orig := isTTYFn
		isTTYFn = func() bool { return true }
		t.Cleanup(func() { isTTYFn = orig })

		got, err := EvaluateCondition(Condition{Type: "no_tty"})
		assertBool(t, err, got, false)
	})
}

func TestEvaluateCondition_AnyOf(t *testing.T) {
	t.Run("true when at least one child is true", func(t *testing.T) {
		t.Setenv("B", "1")
		os.Unsetenv("A")
		got, err := EvaluateCondition(Condition{
			Type: "anyOf",
			Conditions: []Condition{
				{Type: "env_set", Name: "A"},
				{Type: "env_set", Name: "B"},
			},
		})
		assertBool(t, err, got, true)
	})

	t.Run("false when all children are false", func(t *testing.T) {
		os.Unsetenv("A")
		os.Unsetenv("B")
		got, err := EvaluateCondition(Condition{
			Type: "anyOf",
			Conditions: []Condition{
				{Type: "env_set", Name: "A"},
				{Type: "env_set", Name: "B"},
			},
		})
		assertBool(t, err, got, false)
	})

	t.Run("false for an empty condition list", func(t *testing.T) {
		got, err := EvaluateCondition(Condition{Type: "anyOf", Conditions: []Condition{}})
		assertBool(t, err, got, false)
	})

	t.Run("true when a later child is satisfied but an earlier one is not", func(t *testing.T) {
		t.Setenv("A", "")
		t.Setenv("B", "1")
		got, err := EvaluateCondition(Condition{
			Type: "anyOf",
			Conditions: []Condition{
				{Type: "env_set", Name: "A"},
				{Type: "env_set", Name: "B"},
			},
		})
		assertBool(t, err, got, true)
	})
}

func TestEvaluateCondition_AllOf(t *testing.T) {
	t.Run("true when all children are true", func(t *testing.T) {
		t.Setenv("A", "1")
		t.Setenv("B", "1")
		got, err := EvaluateCondition(Condition{
			Type: "allOf",
			Conditions: []Condition{
				{Type: "env_set", Name: "A"},
				{Type: "env_set", Name: "B"},
			},
		})
		assertBool(t, err, got, true)
	})

	t.Run("false when any child is false", func(t *testing.T) {
		t.Setenv("A", "1")
		os.Unsetenv("B")
		got, err := EvaluateCondition(Condition{
			Type: "allOf",
			Conditions: []Condition{
				{Type: "env_set", Name: "A"},
				{Type: "env_set", Name: "B"},
			},
		})
		assertBool(t, err, got, false)
	})

	t.Run("true for an empty condition list (vacuous truth)", func(t *testing.T) {
		got, err := EvaluateCondition(Condition{Type: "allOf", Conditions: []Condition{}})
		assertBool(t, err, got, true)
	})

	t.Run("false when a later child is unsatisfied", func(t *testing.T) {
		t.Setenv("A", "1")
		t.Setenv("B", "")
		got, err := EvaluateCondition(Condition{
			Type: "allOf",
			Conditions: []Condition{
				{Type: "env_set", Name: "A"},
				{Type: "env_set", Name: "B"},
			},
		})
		assertBool(t, err, got, false)
	})
}

func TestEvaluateCondition_Nesting(t *testing.T) {
	coworkCond := Condition{
		Type: "allOf",
		Conditions: []Condition{
			{Type: "env_set", Name: "CLAUDE_CODE_IS_COWORK"},
			{
				Type: "anyOf",
				Conditions: []Condition{
					{Type: "env_set", Name: "CLAUDECODE"},
					{Type: "env_set", Name: "CLAUDE_CODE"},
				},
			},
		},
	}

	t.Run("evaluates allOf(env_set, anyOf(...)) tree", func(t *testing.T) {
		t.Setenv("CLAUDE_CODE_IS_COWORK", "1")
		t.Setenv("CLAUDECODE", "1")
		got, err := EvaluateCondition(coworkCond)
		assertBool(t, err, got, true)
	})

	t.Run("false when the nested anyOf has no satisfied child", func(t *testing.T) {
		t.Setenv("CLAUDE_CODE_IS_COWORK", "1")
		t.Setenv("CLAUDECODE", "")
		t.Setenv("CLAUDE_CODE", "")
		got, err := EvaluateCondition(coworkCond)
		assertBool(t, err, got, false)
	})
}

func TestEvaluateCondition_UnknownType(t *testing.T) {
	t.Run("returns false for an unrecognized type", func(t *testing.T) {
		got, err := EvaluateCondition(Condition{Type: "not_a_real_type"})
		assertBool(t, err, got, false)
	})
}

func assertBool(t *testing.T, err error, got bool, want bool) {
	t.Helper()
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if got != want {
		t.Errorf("got %v, want %v", got, want)
	}
}
